lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -2353,3 +2358,139 @@ bool ReplaceOpenCLBuiltinPass::replaceFract(Function &F, int vec_size) {
+        // add = add sext_op0 sext_op1
+        // clamp = clamp(add, min, max)
+        // result = trunc clamp
+        unsigned extended_bits = static_cast<unsigned>(bitwidth << 1);
+        // The clamp values are the signed min and max of the original bitwidth
+        // sign extended to the extended bitwidth.
+        Constant *scalar_min = ConstantInt::get(
+            Call->getContext(),
+            APInt::getSignedMinValue(bitwidth).sext(extended_bits));
+        Constant *scalar_max = ConstantInt::get(
+            Call->getContext(),
+            APInt::getSignedMaxValue(bitwidth).sext(extended_bits));
+        Constant *min = scalar_min;
+        Constant *max = scalar_max;
+        if (auto vec_ty = dyn_cast<VectorType>(ty)) {
+          min = ConstantVector::getSplat(vec_ty->getElementCount(), min);
+          max = ConstantVector::getSplat(vec_ty->getElementCount(), max);
+        }
+        Type *extended_scalar_ty =
+            IntegerType::get(Call->getContext(), extended_bits);
+        Type *extended_ty = extended_scalar_ty;
+        if (auto vec_ty = dyn_cast<VectorType>(ty)) {
+          extended_ty =
+              VectorType::get(extended_scalar_ty, vec_ty->getElementCount());
+        }
+        auto sext_op0 =
+            CastInst::Create(Instruction::SExt, op0, extended_ty, "", Call);
+        auto sext_op1 =
+            CastInst::Create(Instruction::SExt, op1, extended_ty, "", Call);
+        // Add the nsw flag since we know no overflow can occur.
+        auto add = BinaryOperator::CreateNSW(Instruction::Add, sext_op0,
+                                             sext_op1, "", Call);
+        FunctionType *func_ty = FunctionType::get(
+            extended_ty, {extended_ty, extended_ty, extended_ty}, false);
+
+        // Don't use the type in GetMangledFunctionName to ensure we get
+        // signed parameters.
+        std::string sclamp_name = Builtins::GetMangledFunctionName("clamp");
+        uint32_t vec_width = 1;
+        if (auto vec_ty = dyn_cast<VectorType>(ty)) {
+          vec_width = vec_ty->getElementCount().getKnownMinValue();
+        }
+        if (extended_bits == 32) {
+          if (vec_width == 1) {
+            sclamp_name += "iii";
+          } else {
+            sclamp_name += "Dv" + std::to_string(vec_width) + "_iS_S_";
+          }
+        } else {
+          if (vec_width == 1) {
+            sclamp_name += "sss";
+          } else {
+            sclamp_name += "Dv" + std::to_string(vec_width) + "_sS_S_";
+          }
+        }
+        auto sclamp_callee = module->getOrInsertFunction(sclamp_name, func_ty);
+        auto clamp = CallInst::Create(sclamp_callee, {add, min, max}, "", Call);
+        result = CastInst::Create(Instruction::Trunc, clamp, ty, "", Call);
+      } else {
+        // Pseudo-code:
+        // c = a + b;
+        // if (a < 0 && b < 0)
+        //   c = c > 0 ? min : c;
+        // else if (a >= 0 && b >= 0)
+        //   c = c < 0 ? max : c;
```

Comment:
I think you could simplify this a bit to something like:
```
// c = a + b;
// if (b < 0)
//   c = c > a ? min : c;
// else
//   c = c < a ? max : c;
```

Reply:
Good catch, much simpler.